add add_loggers implementation using PNP

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92058
* __->__ #91639

Summary:

This PR reimplements the old `add_loggers(name_a, model_a, name_b, model_b)`
API in a single-model API style, similar to PNP. This allows for memory
efficiency savings of not having to load two models.

Test plan:

```
python test/test_quantization.py -k NShadows.test_add_loggers
```

vkuzo committed:
[in-progress] add add_loggers implementation using PNP

Summary:

This PR reimplements the old `add_loggers(name_a, model_a, name_b, model_b)`
API in a single-model API style, similar to PNP. This allows for memory
efficiency savings of not having to load two models.

TODO(later): write up in more detail, just saving work fow now.

Test plan:

```
python test/test_quantization.py -k NShadows.test_add_loggers
```

[ghstack-poisoned]

vkuzo committed:
Update on "[in-progress] add add_loggers implementation using PNP"


Summary:

This PR reimplements the old `add_loggers(name_a, model_a, name_b, model_b)`
API in a single-model API style, similar to PNP. This allows for memory
efficiency savings of not having to load two models.

TODO(later): write up in more detail, just saving work fow now.

Test plan:

```
python test/test_quantization.py -k NShadows.test_add_loggers
```

[ghstack-poisoned]

vkuzo committed:
Update on "[in-progress] add add_loggers implementation using PNP"


Summary:

This PR reimplements the old `add_loggers(name_a, model_a, name_b, model_b)`
API in a single-model API style, similar to PNP. This allows for memory
efficiency savings of not having to load two models.

TODO(later): write up in more detail, just saving work fow now.

Test plan:

```
python test/test_quantization.py -k NShadows.test_add_loggers
```

[ghstack-poisoned]

vkuzo committed:
Update on "[in-progress] add add_loggers implementation using PNP"


Summary:

This PR reimplements the old `add_loggers(name_a, model_a, name_b, model_b)`
API in a single-model API style, similar to PNP. This allows for memory
efficiency savings of not having to load two models.

TODO(later): write up in more detail, just saving work fow now.

Test plan:

```
python test/test_quantization.py -k NShadows.test_add_loggers
```

[ghstack-poisoned]

vkuzo committed:
Update on "[in-progress] add add_loggers implementation using PNP"


Summary:

This PR reimplements the old `add_loggers(name_a, model_a, name_b, model_b)`
API in a single-model API style, similar to PNP. This allows for memory
efficiency savings of not having to load two models.

TODO(later): write up in more detail, just saving work fow now.

Test plan:

```
python test/test_quantization.py -k NShadows.test_add_loggers
```

[ghstack-poisoned]

vkuzo committed:
Update on "[in-progress] add add_loggers implementation using PNP"


Summary:

This PR reimplements the old `add_loggers(name_a, model_a, name_b, model_b)`
API in a single-model API style, similar to PNP. This allows for memory
efficiency savings of not having to load two models.

TODO(later): write up in more detail, just saving work fow now.

Test plan:

```
python test/test_quantization.py -k NShadows.test_add_loggers
```

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_numeric_suite_fx.py:
```diff
@@ -2565,6 +2567,183 @@ def custom_convert_fn(module, to_print):
+        self._test_add_loggers_impl(m, example_input, qconfig_mapping)
+
+        if False:
+            # TODO(before land): remove this
```
is this comment still valid?

vkuzo replied:
yes, this is just some extra checks which I'll remove before landing.  There isn't a good way to do these tests without referencing the old NS tools, so it doesn't make sense to add a test since the old tools will be deleted in the future.

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -662,6 +663,247 @@ def handle_subgraph(
+
+        if insert_submodule_copy:
+            match_name = first_node.name
+            handle_subgraph(
```
what does handle_subgraph do? I think it might be helpful to add a comment or change the name to be more specific

vkuzo replied:
sure

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -662,6 +663,247 @@ def handle_subgraph(
+            expected_shadow_name = f"shadow_wrapper_{cur_subgraph_idx}_1"
+            new_shadow_mod = None
+            for maybe_shadow_mod in model.graph.nodes:
+                if maybe_shadow_mod.name == expected_shadow_name:
```
I remember node name does not have a meaning in fx graph and they can be changed without changing the meaning of the graph/program, should we rely on that?

jerryzh168 replied:
yeah the name is still not part of the program definition though.. I'm wondering if we can store the shadow_wrapper module node when we create it? is it created in `handle_subgraph`?

vkuzo replied:
hmm, so these are attribute names actually (all the wrappers and loggers are modules), and the node name is currently taken from attribute name.  If I understand what you are saying, it's that node name may not match the attribute name.  That makes sense.

I think there are two easy options to improve this:
1. instead of checking node.name, check the attribute name of the module (they are the same, but this check is guaranteed).  It would still be string matching though
2. improve on 1 by storing a map of module names (removes the string matching)

are you ok with (1)?  If not, (2) is fine too.

jerryzh168 replied:
yeah (1) sounds good to me

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -662,6 +663,247 @@ def handle_subgraph(
+                    attr_name, args=(last_node,), kwargs={})
+                insertion_point = logger
+
+            cur_node_orig = first_node
```
nit: can we have a summary for what each code block is doing so that it's easier to understand

vkuzo replied:
I made it slightly clearer, let me know if you think we need more comments, happy to expand further

jerryzh168 replied:
thanks, looks good!

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -662,6 +663,247 @@ def handle_subgraph(
+                    )
+                    if first_node_copy is None:
+                        first_node_copy = cur_node_copy
+                cur_node_orig = list(cur_node_orig.users.keys())[0]
```
IIUC, the subgraph is a sequence of ops, should we add an assert here to say that we expect number of users will be 1 here?

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -662,6 +663,247 @@ def handle_subgraph(
+            cur_node_orig = first_node
+            cur_node_copy = None
+            first_node_copy = None
+            while cur_node_orig in subgraph_to_use:
```
optional nit: I'm wondering if we can make this a for loop instead of a while loop, I feel that might be slightly clearer

also maybe in the future, subgraph can be represented as real Graph, and we can use the official api to do the copy: https://github.com/pytorch/pytorch/blob/master/torch/fx/graph.py#L760

vkuzo replied:
hmm, currently we do iteration by walking the graph:

```
cur_node_orig = list(cur_node_orig.users.keys())[0]
```

is a for loop safe?  It would be safe if the quantization matching logic is guaranteed to return ordered nodes.  I feel like it's true, but hoping not to depend on it...walking the graph should be safer in more cases.  Thoughts?

jerryzh168 replied:
oh I see, quantization find_matches is going to return a node arranged as a pattern that is the same as https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/backend_config/README.md#advanced-pattern-specification

I think it's fine, we can refactor this later after we have the operator pattern design in quantization

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -662,6 +663,247 @@ def handle_subgraph(
+        cur_shadow_input.kwargs = tree_map(
+            maybe_remap_node_to_shadow, cur_shadow_input.kwargs)
+
+        cur_subgraph_idx += 1
```
looks like this is unused in this loop?

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -662,6 +663,247 @@ def handle_subgraph(
+
+    model.recompile()
+
+    # Now, we go from
```
I think I understand what `orig_first_node_to_shadow_in_node` does after reading the second code block, maybe it would be helpful to give an example of the content of `orig_first_node_to_shadow_in_node` and `orig_first_node_to_shadow_out_node` here, I think that would make it easier to understand both the previous and following code block

vkuzo replied:
added!

jerryzh168 reviewed (approved):
LGTM, had a few optional inline comments, please also fix the lint comments before landing

vkuzo commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -662,6 +663,247 @@ def handle_subgraph(
+    # need to record original list because we will mutate the graph as we go
+    orig_nodes = list(model.graph.nodes)  # type: ignore[union-attr, arg-type]
+    cur_subgraph_idx = 0
+    orig_first_node_to_shadow_in_node = {}
+    orig_first_node_to_shadow_out_node = {}
```
will do

vkuzo committed:
Update on "add add_loggers implementation using PNP"


Summary:

This PR reimplements the old `add_loggers(name_a, model_a, name_b, model_b)`
API in a single-model API style, similar to PNP. This allows for memory
efficiency savings of not having to load two models.

Test plan:

```
python test/test_quantization.py -k NShadows.test_add_loggers
```

[ghstack-poisoned]
